### ISSUE TITLE
Command framework problems

### ISSUE THREAD

**Issue by InfRandomness** [l7ssha/nyxx#44]
Hello, I'm having issues with commands in my bot, I wanted to create a command usable by the administrator of the server to delete everything on the server, the problem I'm having is that my for loop inside my command either work and make one iteration before stopping, either do not work at all, or my command simply is not called at all. Working totally randomly. The code :
```
@Command('deleteall', aliases: ['da'])
Future<void> deleteAll(CommandContext ctx) async {
  Restrict(
    userPermissions: [PermissionsConstants.administrator],
    botPermissions: [PermissionsConstants.administrator]
  );
  
  var arguments = ctx.message.toString().split(' ');
  switch (arguments[1]) {
    case '*':
      var message = await ctx.reply(content: 'En cours');
      for (var value in ctx.guild.channels.values) {
        sleep(const Duration(seconds: 2));
        await value
            .delete(
                auditReason:
                    'Asked by ${ctx.author.username} who executed the command deleteall *')
            .catchError(throw ("PermissionError"));
      }
  }

  for (var role in ctx.guild.roles.values) {
    await role.delete(
        auditReason:
            'Asked by ${ctx.author.username} who executed the command deleteall *');
  }
}
```
Thanks in advance for the answers.

**Comment by l7ssha** [comment]
Hi, I can't reproduce your problem. I had to fix your code because you have few mistakes in there.
`Restrict` is an annotation and should be used same as `Command`:
```dart
@Restrict(
    userPermissions: [PermissionsConstants.administrator],
    botPermissions: [PermissionsConstants.administrator]
);
Future<void> deleteAll(CommandContext ctx) async { }
```
You shouldn't use `sleep` function in async context. 
`message` variable isn't used. 
Also check if bot has permissions to perform certain action.

**Comment by InfRandomness** [comment]
Oh yeah, thanks for fixing the code.. I am not understanding the documentation very well.. For the sleep, it was to test if the problem was coming from the rate limit of Discord but it was not, and I forgot to remove it. The bot has all the permissions on the server as he have a role with the administrator permission on top of every other roles.

**Comment by InfRandomness** [comment]
Just modified the code to your recommendations and still does not work, still I discovered that sometimes does delete of the channel, but then stops, all the other times it still does not work at all.

Here's what i'm up with on code level

```
@Restrict(
    userPermissions: [PermissionsConstants.administrator],
    botPermissions: [PermissionsConstants.administrator])
@Command('deleteall', aliases: ['da'])
Future<void> deleteAll(CommandContext ctx) async {
  var arguments = ctx.message.toString().split(' ');
  switch (arguments[1]) {
    case '*':
      var message = await ctx.reply(content: 'En cours');
      for (var value in ctx.guild.channels.values) {
        await value
            .delete(
                auditReason:
                    'Asked by ${ctx.author.username} who executed the command deleteall *')
            .catchError(throw ("PermissionError"));
      }
  }

  for (var role in ctx.guild.roles.values) {
    await role.delete(
        auditReason:
            'Asked by ${ctx.author.username} who executed the command deleteall *');
  }
}
```

**Comment by l7ssha** [comment]
@InfRandomness I'll try your code today and see if i can reproduce your bug.

**Comment by InfRandomness** [comment]
alright

**Comment by InfRandomness** [comment]
Any update on this ?

**Comment by l7ssha** [comment]
I can't reproduce your bug, it can be problem with your permissions on server or channels. I created new  guild added bot with administrator perms and tried your code with success.
I used this code to reproduce your bug: 
```dart

@Restrict(
    userPermissions: [PermissionsConstants.administrator],
    botPermissions: [PermissionsConstants.administrator])
@Command('deleteall', aliases: ['da'])
Future<void> deleteAll(CommandContext ctx) async {
  var arguments = ctx.message.toString().split(' ');
  switch (arguments[1]) {
    case '*':
      for (var value in ctx.guild.channels.values) {
        await value
            .delete(
            auditReason:
            'Asked by ${ctx.author.username} who executed the command deleteall *')
            .catchError(throw ("PermissionError"));
      }
  }

  for (var role in ctx.guild.roles.values) {
    await role.delete(
        auditReason:
        'Asked by ${ctx.author.username} who executed the command deleteall *');
  }
}
```

Closing for now due invalid bug report. Reopen if your get anything new.

**Comment by InfRandomness** [comment]
Hey, I really don't get it. It's still not working as intended. Here's what I tried :

I changed of OS, I changed of computer, install a fresh new dart and all that good stuff, created a new guild and added the bot full permissions over every single channel and even on his role which has the administrator permission (8) + every other checked.

After that being done I just cloned my bot and pasted the code for the actual command that you posted above, and still after that, it only deletes one categories after another (and does not care about the others channels) and still, I have to restart the bot after he deleted one category because otherwise he won't do anymore.
